File: final_assignment_backend_quarkus/src/main/java/finalassignmentbackend/service/OffenseTypeDictService.java
package finalassignmentbackend.service;

import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper;
import com.baomidou.mybatisplus.extension.plugins.pagination.Page;
import finalassignmentbackend.config.websocket.WsAction;
import finalassignmentbackend.entity.OffenseTypeDict;
import finalassignmentbackend.entity.SysRequestHistory;
import finalassignmentbackend.mapper.OffenseTypeDictMapper;
import finalassignmentbackend.mapper.SysRequestHistoryMapper;
import io.quarkus.cache.CacheInvalidate;
import io.quarkus.cache.CacheResult;
import io.quarkus.runtime.annotations.RegisterForReflection;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.transaction.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Objects;
import java.util.logging.Level;
import java.util.logging.Logger;

@ApplicationScoped
@RegisterForReflection
public class OffenseTypeDictService {

    private static final Logger log = Logger.getLogger(OffenseTypeDictService.class.getName());

    @Inject
    OffenseTypeDictMapper offenseTypeDictMapper;

    @Inject
    SysRequestHistoryMapper sysRequestHistoryMapper;

    @Transactional
    @CacheInvalidate(cacheName = "offenseTypeDictCache")
    @WsAction(service = "OffenseTypeDictService", action = "checkAndInsertIdempotency")
    public void checkAndInsertIdempotency(String idempotencyKey, OffenseTypeDict dict, String action) {
        Objects.requireNonNull(dict, "OffenseTypeDict must not be null");
        if (isBlank(idempotencyKey)) {
            throw new IllegalArgumentException("Idempotency key must not be blank");
        }
        SysRequestHistory history = sysRequestHistoryMapper.selectByIdempotencyKey(idempotencyKey);
        if (history != null) {
            throw new RuntimeException("Duplicate offense type dict request detected");
        }
        SysRequestHistory newHistory = buildHistory(idempotencyKey);
        sysRequestHistoryMapper.insert(newHistory);
        newHistory.setBusinessStatus("SUCCESS");
        newHistory.setBusinessId(dict.getTypeId() != null ? dict.getTypeId().longValue() : null);
        newHistory.setRequestParams("PENDING");
        newHistory.setUpdatedAt(LocalDateTime.now());
        sysRequestHistoryMapper.updateById(newHistory);
    }

    @Transactional
    @CacheInvalidate(cacheName = "offenseTypeDictCache")
    public OffenseTypeDict createDict(OffenseTypeDict dict) {
        validateDict(dict);
        offenseTypeDictMapper.insert(dict);
        return dict;
    }

    @Transactional
    @CacheInvalidate(cacheName = "offenseTypeDictCache")
    public OffenseTypeDict updateDict(OffenseTypeDict dict) {
        validateDict(dict);
        requirePositive(dict.getTypeId());
        int rows = offenseTypeDictMapper.updateById(dict);
        if (rows == 0) {
            throw new IllegalStateException("No OffenseTypeDict updated for id=" + dict.getTypeId());
        }
        return dict;
    }

    @Transactional
    @CacheInvalidate(cacheName = "offenseTypeDictCache")
    public void deleteDict(Integer typeId) {
        requirePositive(typeId);
        int rows = offenseTypeDictMapper.deleteById(typeId);
        if (rows == 0) {
            throw new IllegalStateException("No OffenseTypeDict deleted for id=" + typeId);
        }
    }

    @CacheResult(cacheName = "offenseTypeDictCache")
    public OffenseTypeDict findById(Integer typeId) {
        requirePositive(typeId);
        return offenseTypeDictMapper.selectById(typeId);
    }

    @CacheResult(cacheName = "offenseTypeDictCache")
    public List<OffenseTypeDict> findAll() {
        return offenseTypeDictMapper.selectList(null);
    }

    @CacheResult(cacheName = "offenseTypeDictCache")
    public List<OffenseTypeDict> searchByOffenseCodePrefix(String offenseCode, int page, int size) {
        if (isBlank(offenseCode)) {
            return List.of();
        }
        validatePagination(page, size);
        QueryWrapper<OffenseTypeDict> wrapper = new QueryWrapper<>();
        wrapper.likeRight("offense_code", offenseCode);
        return fetchFromDatabase(wrapper, page, size);
    }

    @CacheResult(cacheName = "offenseTypeDictCache")
    public List<OffenseTypeDict> searchByOffenseCodeFuzzy(String offenseCode, int page, int size) {
        if (isBlank(offenseCode)) {
            return List.of();
        }
        validatePagination(page, size);
        QueryWrapper<OffenseTypeDict> wrapper = new QueryWrapper<>();
        wrapper.like("offense_code", offenseCode);
        return fetchFromDatabase(wrapper, page, size);
    }

    @CacheResult(cacheName = "offenseTypeDictCache")
    public List<OffenseTypeDict> searchByOffenseNamePrefix(String offenseName, int page, int size) {
        if (isBlank(offenseName)) {
            return List.of();
        }
        validatePagination(page, size);
        QueryWrapper<OffenseTypeDict> wrapper = new QueryWrapper<>();
        wrapper.likeRight("offense_name", offenseName);
        return fetchFromDatabase(wrapper, page, size);
    }

    @CacheResult(cacheName = "offenseTypeDictCache")
    public List<OffenseTypeDict> searchByOffenseNameFuzzy(String offenseName, int page, int size) {
        if (isBlank(offenseName)) {
            return List.of();
        }
        validatePagination(page, size);
        QueryWrapper<OffenseTypeDict> wrapper = new QueryWrapper<>();
        wrapper.like("offense_name", offenseName);
        return fetchFromDatabase(wrapper, page, size);
    }

    @CacheResult(cacheName = "offenseTypeDictCache")
    public List<OffenseTypeDict> searchByCategory(String category, int page, int size) {
        if (isBlank(category)) {
            return List.of();
        }
        validatePagination(page, size);
        QueryWrapper<OffenseTypeDict> wrapper = new QueryWrapper<>();
        wrapper.eq("category", category);
        return fetchFromDatabase(wrapper, page, size);
    }

    @CacheResult(cacheName = "offenseTypeDictCache")
    public List<OffenseTypeDict> searchBySeverityLevel(String severityLevel, int page, int size) {
        if (isBlank(severityLevel)) {
            return List.of();
        }
        validatePagination(page, size);
        QueryWrapper<OffenseTypeDict> wrapper = new QueryWrapper<>();
        wrapper.eq("severity_level", severityLevel);
        return fetchFromDatabase(wrapper, page, size);
    }

    @CacheResult(cacheName = "offenseTypeDictCache")
    public List<OffenseTypeDict> searchByStatus(String status, int page, int size) {
        if (isBlank(status)) {
            return List.of();
        }
        validatePagination(page, size);
        QueryWrapper<OffenseTypeDict> wrapper = new QueryWrapper<>();
        wrapper.eq("status", status);
        return fetchFromDatabase(wrapper, page, size);
    }

    @CacheResult(cacheName = "offenseTypeDictCache")
    public List<OffenseTypeDict> searchByStandardFineAmountRange(double minAmount, double maxAmount, int page, int size) {
        validatePagination(page, size);
        if (minAmount > maxAmount) {
            return List.of();
        }
        QueryWrapper<OffenseTypeDict> wrapper = new QueryWrapper<>();
        wrapper.between("standard_fine_amount", minAmount, maxAmount);
        return fetchFromDatabase(wrapper, page, size);
    }

    @CacheResult(cacheName = "offenseTypeDictCache")
    public List<OffenseTypeDict> searchByDeductedPointsRange(int minPoints, int maxPoints, int page, int size) {
        validatePagination(page, size);
        if (minPoints > maxPoints) {
            return List.of();
        }
        QueryWrapper<OffenseTypeDict> wrapper = new QueryWrapper<>();
        wrapper.between("deducted_points", minPoints, maxPoints);
        return fetchFromDatabase(wrapper, page, size);
    }

    public boolean shouldSkipProcessing(String idempotencyKey) {
        SysRequestHistory history = sysRequestHistoryMapper.selectByIdempotencyKey(idempotencyKey);
        return history != null
                && "SUCCESS".equalsIgnoreCase(history.getBusinessStatus())
                && "DONE".equalsIgnoreCase(history.getRequestParams());
    }

    public void markHistorySuccess(String idempotencyKey, Integer typeId) {
        SysRequestHistory history = sysRequestHistoryMapper.selectByIdempotencyKey(idempotencyKey);
        if (history == null) {
            log.log(Level.WARNING, "Cannot mark success for missing idempotency key {0}", idempotencyKey);
            return;
        }
        history.setBusinessStatus("SUCCESS");
        history.setBusinessId(typeId != null ? typeId.longValue() : null);
        history.setRequestParams("DONE");
        history.setUpdatedAt(LocalDateTime.now());
        sysRequestHistoryMapper.updateById(history);
    }

    public void markHistoryFailure(String idempotencyKey, String reason) {
        SysRequestHistory history = sysRequestHistoryMapper.selectByIdempotencyKey(idempotencyKey);
        if (history == null) {
            log.log(Level.WARNING, "Cannot mark failure for missing idempotency key {0}", idempotencyKey);
            return;
        }
        history.setBusinessStatus("FAILED");
        history.setRequestParams(truncate(reason));
        history.setUpdatedAt(LocalDateTime.now());
        sysRequestHistoryMapper.updateById(history);
    }

    private SysRequestHistory buildHistory(String key) {
        SysRequestHistory history = new SysRequestHistory();
        history.setIdempotencyKey(key);
        history.setBusinessStatus("PROCESSING");
        history.setCreatedAt(LocalDateTime.now());
        history.setUpdatedAt(LocalDateTime.now());
        return history;
    }

    private List<OffenseTypeDict> fetchFromDatabase(QueryWrapper<OffenseTypeDict> wrapper, int page, int size) {
        Page<OffenseTypeDict> mpPage = new Page<>(Math.max(page, 1), Math.max(size, 1));
        offenseTypeDictMapper.selectPage(mpPage, wrapper);
        return mpPage.getRecords();
    }

    private void validateDict(OffenseTypeDict dict) {
        if (dict == null) {
            throw new IllegalArgumentException("OffenseTypeDict must not be null");
        }
        if (dict.getOffenseCode() == null || dict.getOffenseCode().isBlank()) {
            throw new IllegalArgumentException("Offense code must not be blank");
        }
        if (dict.getOffenseName() == null || dict.getOffenseName().isBlank()) {
            throw new IllegalArgumentException("Offense name must not be blank");
        }
        if (dict.getStatus() == null || dict.getStatus().isBlank()) {
            dict.setStatus("Active");
        }
        if (dict.getCreatedAt() == null) {
            dict.setCreatedAt(LocalDateTime.now());
        }
        if (dict.getUpdatedAt() == null) {
            dict.setUpdatedAt(LocalDateTime.now());
        }
    }

    private void validatePagination(int page, int size) {
        if (page < 1 || size < 1) {
            throw new IllegalArgumentException("Page must be >= 1 and size must be >= 1");
        }
    }

    private void requirePositive(Number number) {
        if (number == null || number.longValue() <= 0) {
            throw new IllegalArgumentException("Type ID must be greater than zero");
        }
    }

    private boolean isBlank(String value) {
        return value == null || value.trim().isEmpty();
    }

    private String truncate(String value) {
        if (value == null) {
            return null;
        }
        return value.length() <= 500 ? value : value.substring(0, 500);
    }
}

File: final_assignment_backend_go/project/internal/service/offense_type_dict_service.go
package service

import (
	"errors"
	"strings"
	"time"

	"final_assignment_backend_go/project/internal/domain"
	"final_assignment_backend_go/project/internal/repo"

	"gorm.io/gorm"
)

// OffenseTypeDictService 提供违法类型字典的业务逻辑，对齐 Spring 的
// com.tutict.finalassignmentbackend.service.offense.OffenseTypeDictService。
type OffenseTypeDictService struct {
	repo *repo.OffenseTypeDictRepo
}

func NewOffenseTypeDictService(dictRepo *repo.OffenseTypeDictRepo) *OffenseTypeDictService {
	return &OffenseTypeDictService{repo: dictRepo}
}

func (s *OffenseTypeDictService) DB() *gorm.DB { return s.repo.DB() }

// ErrOffenseTypeDuplicate 对齐 Spring 的重复幂等键拒绝（HTTP 208）。
var ErrOffenseTypeDuplicate = errors.New("duplicate offense type request")

// CheckAndInsertIdempotency 幂等检查 + 落库，沿用 Go 后端统一的幂等账本。
func (s *OffenseTypeDictService) CheckAndInsertIdempotency(key string, dict *domain.OffenseTypeDict, operation string) error {
	if err := checkIdempotency(key, "offense_type:"+operation); err != nil {
		return ErrOffenseTypeDuplicate
	}
	if strings.EqualFold(operation, "create") {
		_, err := s.CreateDict(dict)
		return err
	}
	return s.UpdateDict(dict)
}

// CreateDict 创建违法类型，缺省补齐状态与时间戳。
func (s *OffenseTypeDictService) CreateDict(dict *domain.OffenseTypeDict) (*domain.OffenseTypeDict, error) {
	if dict == nil {
		return nil, errors.New("offense type must not be null")
	}
	if strings.TrimSpace(dict.OffenseCode) == "" {
		return nil, errors.New("offenseCode must not be blank")
	}
	if strings.TrimSpace(dict.OffenseName) == "" {
		return nil, errors.New("offenseName must not be blank")
	}
	now := time.Now()
	if dict.CreatedAt == nil {
		dict.CreatedAt = &now
	}
	if dict.UpdatedAt == nil {
		dict.UpdatedAt = &now
	}
	if strings.TrimSpace(dict.Status) == "" {
		dict.Status = "Active"
	}
	if err := s.repo.Create(dict); err != nil {
		return nil, err
	}
	return dict, nil
}

// UpdateDict 更新违法类型。
func (s *OffenseTypeDictService) UpdateDict(dict *domain.OffenseTypeDict) error {
	if dict == nil || dict.TypeID <= 0 {
		return errors.New("type ID must be greater than zero")
	}
	now := time.Now()
	dict.UpdatedAt = &now
	return s.DB().Save(dict).Error
}

// DeleteDict 删除违法类型（软删除）。
func (s *OffenseTypeDictService) DeleteDict(typeID int) error {
	return s.DB().Where("type_id = ?", typeID).Delete(&domain.OffenseTypeDict{}).Error
}

// FindByID 按主键查询违法类型。
func (s *OffenseTypeDictService) FindByID(typeID int) (*domain.OffenseTypeDict, error) {
	var dict domain.OffenseTypeDict
	err := s.DB().Where("type_id = ?", typeID).First(&dict).Error
	if err != nil {
		return nil, err
	}
	return &dict, nil
}

// FindAll 查询全部违法类型。
func (s *OffenseTypeDictService) FindAll() ([]domain.OffenseTypeDict, error) {
	return s.repo.FindAll()
}

// SearchByOffenseCodePrefix 按违法代码前缀搜索。
func (s *OffenseTypeDictService) SearchByOffenseCodePrefix(code string, page int, size int) ([]domain.OffenseTypeDict, error) {
	return s.dictPage(s.DB().Where("offense_code LIKE ?", prefixLike(code)), page, size)
}

// SearchByOffenseCodeFuzzy 按违法代码模糊搜索。
func (s *OffenseTypeDictService) SearchByOffenseCodeFuzzy(code string, page int, size int) ([]domain.OffenseTypeDict, error) {
	return s.dictPage(s.DB().Where("offense_code LIKE ?", like(code)), page, size)
}

// SearchByOffenseNamePrefix 按违法名称前缀搜索。
func (s *OffenseTypeDictService) SearchByOffenseNamePrefix(name string, page int, size int) ([]domain.OffenseTypeDict, error) {
	return s.dictPage(s.DB().Where("offense_name LIKE ?", prefixLike(name)), page, size)
}

// SearchByOffenseNameFuzzy 按违法名称模糊搜索。
func (s *OffenseTypeDictService) SearchByOffenseNameFuzzy(name string, page int, size int) ([]domain.OffenseTypeDict, error) {
	return s.dictPage(s.DB().Where("offense_name LIKE ?", like(name)), page, size)
}

// SearchByCategory 按违法类别搜索。
func (s *OffenseTypeDictService) SearchByCategory(category string, page int, size int) ([]domain.OffenseTypeDict, error) {
	return s.dictPage(s.DB().Where("category = ?", strings.TrimSpace(category)), page, size)
}

// SearchBySeverityLevel 按严重程度搜索。
func (s *OffenseTypeDictService) SearchBySeverityLevel(severity string, page int, size int) ([]domain.OffenseTypeDict, error) {
	return s.dictPage(s.DB().Where("severity_level = ?", strings.TrimSpace(severity)), page, size)
}

// SearchByStatus 按状态搜索。
func (s *OffenseTypeDictService) SearchByStatus(status string, page int, size int) ([]domain.OffenseTypeDict, error) {
	return s.dictPage(s.DB().Where("status = ?", strings.TrimSpace(status)), page, size)
}

// SearchByStandardFineAmountRange 按标准罚款金额范围搜索。
func (s *OffenseTypeDictService) SearchByStandardFineAmountRange(minAmount float64, maxAmount float64, page int, size int) ([]domain.OffenseTypeDict, error) {
	return s.dictPage(s.DB().Where("standard_fine_amount BETWEEN ? AND ?", minAmount, maxAmount), page, size)
}

// SearchByDeductedPointsRange 按扣分范围搜索。
func (s *OffenseTypeDictService) SearchByDeductedPointsRange(minPoints int, maxPoints int, page int, size int) ([]domain.OffenseTypeDict, error) {
	return s.dictPage(s.DB().Where("deducted_points BETWEEN ? AND ?", minPoints, maxPoints), page, size)
}

func (s *OffenseTypeDictService) dictPage(query *gorm.DB, page int, size int) ([]domain.OffenseTypeDict, error) {
	offset, limit := pageBounds(page, size)
	var dicts []domain.OffenseTypeDict
	err := query.Order("type_id").Offset(offset).Limit(limit).Find(&dicts).Error
	return dicts, err
}
